test/micro/org/openjdk/bench/jdk/classfile/RebuildMethodBodies.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+package org.openjdk.bench.jdk.classfile;
+
+import java.io.IOException;
+import java.net.URI;
+import java.nio.file.FileSystems;
+import java.nio.file.Files;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import jdk.internal.classfile.ClassModel;
+import jdk.internal.classfile.Classfile;
+import jdk.internal.classfile.ClassTransform;
+import jdk.internal.classfile.instruction.*;
+import org.openjdk.jmh.annotations.*;
+
+@BenchmarkMode(Mode.Throughput)
+@State(Scope.Benchmark)
+@Fork(value = 1, jvmArgsAppend = {
+        "--enable-preview",
+        "--add-exports", "java.base/jdk.internal.classfile=ALL-UNNAMED",
+        "--add-exports", "java.base/jdk.internal.classfile.constantpool=ALL-UNNAMED",
+        "--add-exports", "java.base/jdk.internal.classfile.instruction=ALL-UNNAMED"})
+@Warmup(iterations = 2)
+@Measurement(iterations = 4)
+public class RebuildMethodBodies {
+
+    List<ClassModel> shared, unshared;
+    Iterator<ClassModel> it1, it2;
+
+    @Setup(Level.Trial)
+    public void setup() throws IOException {
+        shared = new LinkedList<>();
+        unshared = new LinkedList<>();
```

Comment:
LinkedList should be replaced by ArrayList, as:
1. LinkedList's node wrapper around each object is too heavy, ArrayList has smaller wrapper
2. LinkedList iteration needs to follow links while ArrayList access can be machine optimized
3. ArrayList addition is amortized O(1), not really worse than LinkedList additions.

Reply:
I have to side with @liach here: `LinkedList` iteration is significantly slower than `ArrayList` even though the computational complexity is identical. Often by an integer faction. This is due to the sparseness of the linked list data structure on heap and the pointer chasing that entails.  

For minimum overhead of iteration over an immutable list then turning the list into an immutable via `List.copyOf` might be preferable due the JVM's ability to optimize over `@Stable` arrays.

Adhoc [JMH benchmark](https://gist.github.com/cl4es/1f21812241c47f32a9deaffcc996e8b3):

```
Benchmark                            Mode  Cnt    Score    Error   Units
ListIteration.iterateArrayList      thrpt   15  188,724 ± 10,903  ops/ms
ListIteration.iterateImmutableList  thrpt   15  230,901 ±  6,513  ops/ms
ListIteration.iterateLinkedList     thrpt   15   58,289 ±  0,497  ops/ms
```

Is it important to fix this? Perhaps not, but in a microbenchmark like this the fewer cycles we spend on "stuff" that isn't really part of the thing we're testing, the better. Increasingly so as the thing we're testing is getting more and more optimized.

Reply:
Right, fixed, thanks.